v2/datastream-to-spanner/terraform/samples/mysql-sharded-end-to-end/README.md

```diff
@@ -0,0 +1,689 @@
+
+```shell
+#!/bin/bash
+
+# Service account to be granted roles
+SERVICE_ACCOUNT="<YOUR-SERVICE-ACCOUNT>@<YOUR-PROJECT-ID>.iam.gserviceaccount.com"
+
+# Project ID where roles will be granted
+PROJECT_ID="<YOUR-PROJECT-ID>"
+
+# Array of roles to grant
+ROLES=(
+  "roles/<role1>"
+  "roles/<role2>"
+)
+
+# Loop through each role and grant it to the service account
+for ROLE in "${ROLES[@]}"
+do
+  gcloud projects add-iam-policy-binding "$PROJECT_ID" \
+    --member="serviceAccount:${SERVICE_ACCOUNT}" \
+    --role="$ROLE"
+done
+```
+
+### Verifying access in the Terraform service account
+
+#### Using custom role and granular permissions (recommended)
+
+Verify that the custom role is attached to the service account - 
+
+```shell
+gcloud projects get-iam-policy <YOUR-PROJECT-ID>  \
+--flatten="bindings[].members" \
+--format='table(bindings.role)' \
+--filter="bindings.members:<YOUR-SERVICE-ACCOUNT>@<YOUR-PROJECT-ID>.iam.gserviceaccount.com"
+```
+
+Verify that the role has the correct set of permissions
+
+```shell
+gcloud iam roles describe live_migrations_role --project=<YOUR-PROJECT-ID> 
+```
+
+##### Using pre-defined roles
+Once the roles are added, run the following command to verify them -
+
+```shell
+gcloud projects get-iam-policy <YOUR-PROJECT-ID>  \
+--flatten="bindings[].members" \
+--format='table(bindings.role)' \
+--filter="bindings.members:<YOUR-SERVICE-ACCOUNT>@<YOUR-PROJECT-ID>.iam.gserviceaccount.com"
+```
+
+Sample output -
+
+```shell
+ROLE
+roles/dataflow.admin
+roles/datastream.admin
+roles/iam.securityAdmin
+roles/iam.serviceAccountUser
+roles/pubsub.admin
+roles/storage.admin
+roles/viewer
+```
+
+### Impersonating the Terraform service account
+
+#### Using GCE VM instance (recommended)
+
+A GCE VM created using the service account setup above will automatically
+use the service account for all API requests triggered by Terraform. Running
+terraform from such a GCE VM does not require downloading service keys and is
+the recommended approach.
+
+#### Using key file
+
+1. Activate the service account -
+   ```shell
+   gcloud auth activate-service-account <YOUR-SERVICE-ACCOUNT>@<YOUR-PROJECT-ID>.iam.gserviceaccount.com --key-file=path/to/key_file --project=project_id
+   ```
+2. Impersonate service account while fetching the ADC credentials -
+   ```shell
+   gcloud auth application-default login --impersonate-service-account <YOUR-SERVICE-ACCOUNT>@<YOUR-PROJECT-ID>.iam.gserviceaccount.com
+   ```
+
+## Advanced Topics
+
+### Specifying parallelism for Terraform
+
+Terraform limits the number of concurrent operations while walking the graph.
+[The default](https://developer.hashicorp.com/terraform/cli/commands/apply#parallelism-n)
+is 10.
+
+Increasing this value can
+potentially speed up orchestration execution
+when orchestrating a large sharded migration. We strongly recommend against
+setting this value > 20. In most cases, the default value should suffice.
```

Comment:
Is there harm in setting this value higher as a default ? We should aim to for this to run for the large use cases we are aware of out of the box.

Reply:
Terraform command will be executed by the customer and this will be passed as a flag. There is no notion of setting as default here. Hence, we are providing recommendations.

Yes, in my experiments, values above 20 start causing `429` errors from GCP APIs. I think using 10 even for large orchestrations will result in more stable orchestrations, hence the guidance.

Reply:
Sure